cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceUtil.java

```diff
@@ -41,13 +50,58 @@ private AuthEnforceUtil() {
+   * @throws Exception {@link UnauthorizedException} if the given authenticationContext is not authorized to perform
+   * the specified actions on the entity
    */
-  public static void enforce(AuthorizationEnforcer authorizationEnforcer, EntityId entities,
+  public static void enforce(AuthorizationEnforcer authorizationEnforcer, Object[] entities, Class enforceOn,
                              AuthenticationContext authenticationContext, Set<Action> actions) throws Exception {
-    authorizationEnforcer.enforce(entities, authenticationContext.getPrincipal(), actions);
+    authorizationEnforcer.enforce(getEntityId(entities, enforceOn), authenticationContext.getPrincipal(), actions);
+  }
+
+  private static EntityId getEntityId(Object[] entities, Class enforceOn) {
+    if (entities.length == 1 && entities[0] instanceof EntityId) {
+      // If EntityId was passed in then the size of the array should be one
+      return (EntityId) entities[0];
+    } else {
+      // If the size of the array is more than one we expect all the specified parameters to be String
+      String[] entityParts;
+      try {
+        entityParts = Arrays.copyOf(entities, entities.length, String[].class);
+      } catch (ArrayStoreException e) {
+        // Arrays.copyOf will throw ArrayStoreException if an element copied from original array is not a String
+        throw new IllegalArgumentException(String.format("%s annotation can be used for multiple part entities of " +
+                                                                 "String type only.", AuthEnforce.class.getName()), e);
+      }
+      return EntityIdFactory.getEntityId(entityParts, enforceOn);
+    }
+  }
+
+  private static class EntityIdFactory {
+    public static EntityId getEntityId(String[] entityParts, Class enforceOn) {
+      if (enforceOn.equals(InstanceId.class)) {
+        return new InstanceId(entityParts[0]);
+      } else if (enforceOn.equals(NamespaceId.class)) {
+        return new NamespaceId(entityParts[0]);
+      } else if (enforceOn.equals(ArtifactId.class)) {
+        return new ArtifactId(entityParts[0], entityParts[1], entityParts[2]);
+      } else if (enforceOn.equals(DatasetId.class)) {
+        return new DatasetId(entityParts[0], entityParts[1]);
+      } else if (enforceOn.equals(StreamId.class)) {
+        return new StreamId(entityParts[0], entityParts[1]);
+      } else if (enforceOn.equals(ApplicationId.class)) {
+        return new ApplicationId(entityParts[0], entityParts[1]);
+      } else if (enforceOn.equals(ProgramId.class)) {
+        return new ProgramId(entityParts[0], entityParts[1], entityParts[2], entityParts[3]);
+      } else {
+        // safeguard: this should not happen since we have already verified all that the entity id can be created from
+        // the given parts
+        throw new IllegalArgumentException(String.format("Failed to create an %s from the the given entity parts %s",
```

Comment:
There are other classes that inherit from `EntityId` that are not included above. I think it's better to use reflection (with cached constructor) to do this.

```java
  private static final Map<Class<? extends EntityId>, Constructor<? extends EntityId>> CONS_CACHE =
    new ConcurrentHashMap<>();

  private static EntityId createEntityId(Class<? extends EntityId> entityClass,
                                         Object[] args) throws IllegalAccessException,
                                                               InvocationTargetException,
                                                               InstantiationException,
                                                               NoSuchMethodException {
    Constructor<? extends EntityId> cons = CONS_CACHE.get(entityClass);
    if (cons != null) {
      return cons.newInstance(args);
    }

    cons = findConstructor(entityClass);

    // It's ok to just put. If there are concurrent calls, both of them should end up with the same constructor.
    CONS_CACHE.put(entityClass, cons);
    return cons.newInstance(args);
  }

  private static Constructor<? extends EntityId> findConstructor(Class<? extends EntityId> entityClass) {
    // Find the constructor with all String parameters
    for (Constructor<?> constructor : entityClass.getConstructors()) {
      if (Iterables.all(Arrays.asList(constructor.getParameterTypes()), Predicates.<Class<?>>equalTo(String.class))) {
        return (Constructor<? extends EntityId>) constructor;
      }
    }
    throw new IllegalStateException("Failed to find constructor on " + entityClass.getName()
                                      + " that parameters are only of String type");
  }

```